channel/ledger/channel.go

```diff
@@ -0,0 +1,258 @@
+var ErrDuplicateGuarantee = fmt.Errorf("duplicate guarantee detected")
+
+// DivertToGuarantee deducts g.amount from o.left's balance, and
+// adds g to o.guarantees
+func (o LedgerOutcome) DivertToGuarantee(p Add) (LedgerOutcome, error) {
+	g := p.Guarantee
+
+	if types.Gt(&p.LeftDeposit, &g.amount) {
+		return LedgerOutcome{}, ErrInvalidDeposit
+	}
+
+	if types.Gt(&g.amount, &o.left.amount) {
+		return LedgerOutcome{}, ErrInsufficientFunds
+	}
+
+	o.left.amount.Sub(&o.left.amount, &p.LeftDeposit)
+	rightDeposit := p.RightDeposit()
+	o.right.amount.Sub(&o.right.amount, &rightDeposit)
+
+	_, found := o.guarantees[g.target]
+	if found {
+		return LedgerOutcome{}, ErrDuplicateGuarantee
+	}
+	o.guarantees[g.target] = g
+
+	return o, nil
+}
+
+// Vars stores the turn number and outcome for a state in a consensus channel
+type Vars struct {
+	TurnNum uint64
+	Outcome LedgerOutcome
+}
+
+// SignedVars stores 0-2 signatures for some vars in a consensus channel
+type SignedVars struct {
+	Vars
+	Signatures [2]*state.Signature
+}
+
+// SignedProposal is a proposal with a signature on it
+type SignedProposal struct {
+	state.Signature
+	Proposal interface{}
+}
+
+// Add is a proposal to add a guarantee for the given virtual channel
+// amount is to be deducted from left
+type Add struct {
+	turnNum uint64
+	Guarantee
+	LeftDeposit big.Int
+}
+
+func (a Add) RightDeposit() big.Int {
+	result := big.Int{}
+	result.Sub(&a.amount, &a.LeftDeposit)
+
+	return result
+}
+
+// Remove is a proposal to remove a guarantee from the ledger channel
+// - `paid` funds are added to the right participant's balance
+// - the remainder are returned to the left participant's balance
+type Remove struct {
+	TurnNum uint64
+	VId     types.Destination
+	Paid    types.Funds
+}
+
+var ErrIncorrectTurnNum = fmt.Errorf("incorrect turn number")
+
+// Add updates Vars by including a guarantee, updating balances accordingly
+func (vars Vars) Add(p Add) (Vars, error) {
+	if p.turnNum != vars.TurnNum+1 {
+		return Vars{}, ErrIncorrectTurnNum
+	}
+
+	vars.TurnNum += 1
+
+	o, err := vars.Outcome.DivertToGuarantee(p)
+
+	if err != nil {
+		return Vars{}, err
+	}
+
+	vars.Outcome = o
+	return vars, nil
+}
+
+func (vars Vars) Remove(p Remove) (Vars, error) {
+	panic("UNIMPLEMENTED")
+}
+
+// Add receives a Guarantee, and generates and stores a SignedProposal in
+// the queue, returning the resulting SignedProposal
+func (c *ConsensusChannel) Add(g Guarantee, sk []byte) (SignedProposal, error) {
+	if c.MyIndex != proposerIndex {
+		return SignedProposal{}, fmt.Errorf("only proposer can call Add")
+	}
+
+	vars := c.current.Vars
+	var err error
+
+	for _, p := range c.proposalQueue {
+		vars, err = vars.Add(p.Proposal.(Add))
+		if err != nil {
+			return SignedProposal{}, err
+		}
+	}
+
+	latest := c.proposalQueue[len(c.proposalQueue)-1].Proposal.(Add)
+
+	vars, err = vars.Add(Add{Guarantee: g, turnNum: latest.turnNum + 1})
+	if err != nil {
+		return SignedProposal{}, err
```

Comment:
nit but instead of returning the `err` from `vars.Add` directly it'd be nice to wrap the error.
```suggestion
		return SignedProposal{}, fmt.Errorf("propose could not add new state vars: %w", err)
```